### PR TITLE
Adding support for request body and http headers

### LINKED ISSUE #2379: Lambda Plugin Headers
Hey guys, great work on Kong! I have been evaluating the Lambda plugin for Kong.

During my testing I've found that none of the headers get passed over from Kong. I would have thought that any headers from the original request would be forwarded to lambda through the event? This seems to work for post data, but not headers :(

This would be particularly useful we so can extract the auth headers and http method 👍

### LINKED ISSUE #2469: aws-lambda plugin works incorrect with some content-types
### Summary

aws-lambda plugin works incorrect with some content-types. The plugin works correctly only with application/json, multipart/form-data and application/x-www-form-urlencoded. 
Other content-types/requests are processed incorrectly they are treated as application/x-www-form-urlencoded type, so it's impossible to use the plugin with SOAP for instance or other request with text/binary payload 

### Steps To Reproduce

1.
curl -H "Content-Type: application/xml" -H "Accept: application/json" -X POST 'http://localhost:8000/txb2fms' -d'<x/>' 2>/dev/null | jq
{
  **"<x/>"**: true
}

### Additional Details & Logs

- Kong version 0.10.0
- Operating System - Ubuntu

### PR DESCRIPTION
**Pull request by akh00**
### Summary

Fixed work with application/xml, application/soap+xml and text/* and other (binary) content-types.
Added support for http headers passing to lambda
Changes are backward compatible 

### Full changelog

* schema.lua - added the following Boolean properties 
    forward_request_body = { type = "boolean", default = false, required = true }, -- if the request body should be in separate field
    forward_request_http_headers = { type = "boolean", default = false, required = true }, -- if the http headers should be passed in separate field
    forward_request_http_method = { type = "boolean", default = false, required = true }, -- if the http method name should be passed to lambda
   forward_request_uri = { type = "boolean", default = false, required = true } -- if the request url should be passed to lambda

 * handler.lua - changed accordingly

### Issues resolved

Fix #2379 
Fix #2469

### CODE REVIEW

**Comment by akh00** [comment]
please review

**Comment by akh00** [comment]
I've removed using io library. Also I've manually tested the modification using "echo" lambda. I'm not sure how I can create some automation tests for this ( and I didn't find any existing). However for my use case (replacing existing web service with lambda) soap request body support is crucial. Please review and advise what else should be done

**Comment by p0pr0ck5** [comment]
@akh00 the tests are in `spec/03-plugins/23-aws-lambda/01-access_spec.lua`; however, we're going to be adjusted them a bit to rely on a proper environment (see https://github.com/Mashape/kong/pull/2287)

File: kong/plugins/aws-lambda/handler.lua [hunk 1]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
     body_parameters = Multipart(ngx_req_get_body_data(), content_type):get_all()
-  elseif content_type and string_find(content_type:lower(), "application/json", nil, true) then
+  elseif content_type and string_find(content_type:lower(), "application/json", nil, true)
+         and not conf.request_body then
     body_parameters, err = cjson.decode(ngx_req_get_body_data())
     if err then
       body_parameters = {}
     end
-  else
+  elseif content_type and string_find(content_type:lower(), "application/x-www-form-urlencoded", nil, true) then
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: repeated `:lower()` call are unnecesary. let's find a way to make this function call only one time, if necessary (is it?).

File: kong/plugins/aws-lambda/handler.lua [hunk 2]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
+  local err, body, request
+  local rawBody = ngx_req_get_body_data()
+  if rawBody == nil then return nil end
+
+  if content_type and string_find(content_type:lower(), "application/json", nil, true) then
+    body, err = cjson.decode(rawBody)
+    if err then
+      return nil
+    end
+  elseif content_type and ( string_find(content_type:lower(), "application/xml", nil, true)
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: style: no space between paren and first function param

File: kong/plugins/aws-lambda/handler.lua [hunk 3]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
     body_parameters = public_utils.get_post_args()
+  else
+    body_parameters = {}
   end
 
   return utils.table_merge(ngx_req_get_uri_args(), body_parameters)
 end
 
+local function retrieve_body(content_type)
+  ngx_req_read_body()
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: we've already read the body once we've reach this point; this call is unnecessary.

File: kong/plugins/aws-lambda/handler.lua [hunk 4]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
+  local err, body, request
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: where is the value `request` used?

File: kong/plugins/aws-lambda/handler.lua [hunk 5]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
+  local rawBody = ngx_req_get_body_data()
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: style: primitive object variable shouldn't be camel cased

File: kong/plugins/aws-lambda/handler.lua [hunk 6]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
+  if rawBody == nil then return nil end
+
+  if content_type and string_find(content_type:lower(), "application/json", nil, true) then
+    body, err = cjson.decode(rawBody)
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: does the OpenResty version of `cjson` return a second `err` return val? the upstream version doesn't: https://www.kyne.com.au/~mark/software/lua-cjson-manual.html#_decode

File: kong/plugins/aws-lambda/handler.lua [hunk 7]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
+    if err then
+      return nil
+    end
+  elseif content_type and ( string_find(content_type:lower(), "application/xml", nil, true)
+    or  string_find(content_type:lower(), "application/soap+xml", nil, true)
+    or  string_find(content_type:lower(), "text/", nil, true) )
+  then
+    body = rawBody
+  else
+    body =  ngx_encode_base64( rawBody )
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: there are certainly many content types that are not binary, that would end up Base64 encoded here. perhaps we should invert this logic, and only encode certain known content types, and raw pass everything else?

File: kong/plugins/aws-lambda/handler.lua [hunk 8]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
+  end
+  return body
+end
+
 function AWSLambdaHandler:access(conf)
   AWSLambdaHandler.super.access(self)
 
-  local bodyJson = cjson.encode(retrieve_parameters())
+  local req_headers = ngx_req_get_headers()
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: if these values are not use elsewhere in this function, perhaps we could wrap this chunk in a `do` block to minimize the number of in-play local vars.

File: kong/plugins/aws-lambda/handler.lua [hunk 9]
```diff
@@ -23,28 +27,71 @@ function AWSLambdaHandler:new()
...
+    local body = retrieve_body(content_type)
+    if body ~= nil then
+       body_parameters["request_body"] = body
+    end
+  end
+  if conf.http_headers then
+     body_parameters["http_headers"] = req_headers
+  end
+  if conf.http_method then
+     body_parameters["http_method"] = ngx_req_get_method()
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: there is some inconsistency in this chain of if blocks. here we call a wrapped `ngx.req.get_method()`, but in other cases we assign a value based on a previously-declared variable. can we please keep this style of assignment consistent? preferably, we are not calling into the ngx.* API and assigning variables more than necessary.

File: kong/plugins/aws-lambda/schema.lua [hunk 10]
```diff
@@ -12,6 +12,10 @@ return {
     invocation_type = {type = "string", required = true, default = "RequestResponse", 
                        enum = {"RequestResponse", "Event", "DryRun"}},
     log_type = {type = "string", required = true, default = "Tail", 
-                       enum = {"Tail", "None"}}
+                       enum = {"Tail", "None"}},
+    request_body = {type="boolean", default = false, required = true},
+    http_headers = {type="boolean", default = false, required = true},
+    http_method = {type="boolean", default = false, required = true},
+    request_url = {type="boolean", default = false, required = true}
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/schema.lua]: style: needs a trailing comma

File: kong/plugins/aws-lambda/handler.lua (hunk 3 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: According to documentation https://github.com/openresty/lua-nginx-module#ngxreqread_body 
_If the request body is already read previously by turning on lua_need_request_body or by using other modules, then this function does not run and returns immediately_. 
I'd rather keep function calls independent from the call order and given that it should return without actual reading I'd keep it or if it's absolutely unacceptable move to level up

File: kong/plugins/aws-lambda/handler.lua (hunk 8 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: req_headers is used to read "Content-Type" and to pass to resulting request if it's configured. That's why I'm reading it here

File: kong/plugins/aws-lambda/handler.lua (hunk 9 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: The "method" is used only here I'd rather avoid to call method until it's required.

File: kong/plugins/aws-lambda/handler.lua (hunk 4 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Thank you. Removed

File: kong/plugins/aws-lambda/handler.lua (hunk 5 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Fixed. Thank you

File: kong/plugins/aws-lambda/handler.lua (hunk 6 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: According to https://www.kyne.com.au/~mark/software/lua-cjson-manual.html 
_The cjson.safe module behaves identically to the cjson module, except when errors are encountered during JSON conversion. On error, the cjson_safe.encode and cjson_safe.decode functions will return nil followed by the error message._

File: kong/plugins/aws-lambda/handler.lua (hunk 2 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Fixed. Thanks

File: kong/plugins/aws-lambda/handler.lua (hunk 7 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: It's interesting question. I'd rather say that there is very many binary types and only few text specially in the lambda usage context.

**Comment by akh00** [comment]
I've addressed the review feedback and enabled integration tests with using httpbin.org as the lambda mock. Please review.

**Comment by p0pr0ck5** [comment]
Would like to get a second pair of eyes here, @thibaultcha or @Tieske if you have time, particularly with respect to previous discussion in #2287

File: kong/plugins/aws-lambda/handler.lua (hunk 8 above)
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: it can be preferable to use the `ngx.var` API to retrieve individual headers: https://github.com/openresty/lua-nginx-module#ngxreqget_headers, and that is our style moving forward. also, this doesn't address the comment about closing the scope of these vars :)

File: kong/plugins/aws-lambda/handler.lua [hunk 11]
```diff
@@ -72,14 +123,15 @@ function AWSLambdaHandler:access(conf)
...
   end
-
   -- Trigger request
   local client = http.new()
-  client:connect(host, 443)
   client:set_timeout(conf.timeout)
-  local ok, err = client:ssl_handshake()
-  if not ok then
-    return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
+  client:connect(host, (conf.aws_port or 443))
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: We explicitly discussed in #2287 that this form (mixing application and testing code) is undesired. I think we need to find a different approach here (the above PR will solve this)

File: kong/plugins/aws-lambda/handler.lua (hunk 3 above)
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: I think we should move it up, then, so we aren't needlessly calling multiple times :)

File: kong/plugins/aws-lambda/handler.lua (hunk 11 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: In fact it's not mixing application and test code it's allowing to override default port and host which should be allowed anyway. Along with overriding default path "/2015-03-31/functions/%s/invocations"
There was logic in kong/plugins/aws-lambda/v4.lua that allowed to path port there 
`  local port = tbl.port or (tls and 443 or 80)
`
I only added possibility to pass host as well and make it configurable. I removed configuration for ssl in current version of PR

File: kong/plugins/aws-lambda/handler.lua (hunk 3 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Done

File: kong/plugins/aws-lambda/handler.lua (hunk 8 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Done

File: kong/plugins/aws-lambda/handler.lua (hunk 9 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Fixed

**Comment by p0pr0ck5** [comment]
Thanks @akh00. still would like to get another pair of eyes here ( @bungle your review would be welcome here as well).

**Comment by akh00** [comment]
Please review with new mock approach incorporated

File: kong/plugins/aws-lambda/handler.lua [hunk 12]
```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
...
+    body = ngx_encode_base64(raw_body)
+  end
+  return body
+end
+
 function AWSLambdaHandler:access(conf)
   AWSLambdaHandler.super.access(self)
 
-  local bodyJson = cjson.encode(retrieve_parameters())
+  local content_type = ngx.var.http_content_type
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: We access the `ngx` global a few times during this function. It would be better to cache it:
```lua
local var = ngx.var

local content_type = var.http_content_type

var.request_method
...
```

File: kong/plugins/aws-lambda/handler.lua [hunk 13]
```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
...
+  if conf.http_headers then
+     body_parameters["http_headers"] = ngx_req_get_headers()
+  end
+  if conf.http_method then
+     body_parameters["http_method"] = ngx.var.request_method
+  end
+  if conf.request_uri then
+     body_parameters["request_uri"] = ngx.var.request_uri
+  end
+  local body_json = cjson.encode(body_parameters)
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: This module loads `cjson.safe`, which returns an error if the encoding fails. This call should handle such errors:
```lua
local body_json, err = cjson.encode(body_parameters)
if not body_json then
  ngx.log(ngx.ERR, "[aws-lambda] could not encode JSON ",
                   "body for forwarded request values: ", err)
end
```

Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: style: need a line jump before this assigment too

File: kong/plugins/aws-lambda/handler.lua [hunk 14]
```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
...
+
+  if conf.request_body then
+     body_parameters["request_body"] = retrieve_body(content_type)
+  end
+  if conf.http_headers then
+     body_parameters["http_headers"] = ngx_req_get_headers()
+  end
+  if conf.http_method then
+     body_parameters["http_method"] = ngx.var.request_method
+  end
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: style: those `if` blocks need line jumps in between them as well, for clarity.

File: kong/plugins/aws-lambda/handler.lua [hunk 15]
```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
...
+  elseif content_type and ( string_find(content_type, "application/xml", nil, true)
+    or  string_find(content_type, "application/soap+xml", nil, true)
+    or  string_find(content_type, "text/", nil, true) )
+  then
+    body = raw_body
+  else
+    body = ngx_encode_base64(raw_body)
+  end
+  return body
+end
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: Hmmm why the fallback the base-64 encoding as a fallback option exactly?

File: kong/plugins/aws-lambda/handler.lua [hunk 16]
```diff
@@ -25,32 +24,77 @@ function AWSLambdaHandler:new()
...
+  local content_type = ngx.var.http_content_type
+  if content_type then
+     content_type = content_type:lower()
+  end
+
+  ngx_req_read_body()
 
-  local host = string.format("lambda.%s.amazonaws.com", conf.aws_region)
-  local path = string.format("/2015-03-31/functions/%s/invocations",
+  local body_parameters = retrieve_parameters(content_type, conf)
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: This will need to be refactored once #2503 gets merged, will you mind?

File: kong/plugins/aws-lambda/handler.lua (hunk 15 above)
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: Also, if we already decode the body (when JSON) in `retrieve_parameters`, and include it in the upstream response body, why decoding it again here? Isn't the goal of this `request_body` parameter to be the raw request body?

File: kong/plugins/aws-lambda/handler.lua (hunk 16 above)
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: Also, I'm not sure I feel comfortable with mixing the new values from this PR with the request body, at least at the same root level. For semantics but also due to possible conflicts and overriding a previous body parameter.

I feel like the proper behavior would be:

Instead:
```lua
local upstream_body = {
  request_body_args    = retrieve_parameters(), -- replaced once #2503 gets merged
  request_body_raw     = retrieve_body(content_type),
  request_http_headers = ngx_req_get_headers(),
  request_http_method  = var.request_method,
  request_uri          = var.request_uri,
}
```

Sadly, this would be a breaking change for this plugin, but I do not see a possibility of this being merged as-is, for the reasons described above...

File: kong/plugins/aws-lambda/handler.lua [hunk 17]
```diff
@@ -73,11 +119,10 @@ function AWSLambdaHandler:access(conf)
   if err then
     return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
   end
-
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: Seems like an undesired change?

File: kong/plugins/aws-lambda/handler.lua [hunk 18]
```diff
@@ -60,10 +104,12 @@ function AWSLambdaHandler:access(conf)
       ["X-Amz-Invocation-Type"] = conf.invocation_type,
       ["X-Amx-Log-Type"] = conf.log_type,
       ["Content-Type"] = "application/x-amz-json-1.1",
-      ["Content-Length"] = tostring(#bodyJson)
+      ["Content-Length"] = tostring(#body_json)
     },
-    body = bodyJson,
+    body = body_json,
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: +1 for camel_case, this was a mistake

**Comment by NaveenGoli** [comment]
We need this ASAP, can we get this fixed and merged ASAP?

**Comment by thibaultcha** [comment]
No, sorry. However, you are free to apply this patch manually to your Kong installation (albeit it isn't ready for merging) if you desperately need it :)

File: kong/plugins/aws-lambda/handler.lua (hunk 16 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: What about this? It would be still not breaking change and it still totally makes sense for me.
`  if conf.forward_request_body or conf.forward_http_headers 
     or conf.forward_http_method or conf.forward_request_uri then
     upstream_body = {
       request_body_args = not conf.forward_request_body and retrieve_parameters(content_type)
                           or ngx_req_get_uri_args(), 
       request_body = conf.forward_request_body and retrieve_body(content_type) or nil,
       request_http_headers = conf.forward_http_headers and ngx_req_get_headers() or nil,
       request_http_method = conf.forward_http_method and var.request_method or nil,
       request_uri = conf.forward_request_uri and var.request_uri or nil
     }
  else
     upstream_body = retrieve_parameters(content_type)
  end
`

File: kong/plugins/aws-lambda/handler.lua (hunk 14 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: refactored

File: kong/plugins/aws-lambda/handler.lua (hunk 13 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: I didn't touch this peace of code though you are correct and I've added check

Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Refactored

File: kong/plugins/aws-lambda/handler.lua (hunk 12 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Thanks. Done

File: kong/plugins/aws-lambda/handler.lua (hunk 15 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: The fallback is because most of the other content types are binary.
We didn't decode body when JSON in retrieve parameters in case when request_body is set in order to avoid decoding it twice. I'd say that in case of JSON it makes sense to forward JSON request body as part of JSON wrapper. Though it does not matter for me. Please look at the refactored PR.

**Comment by akh00** [comment]
PR has been changed according changes requested. Please review

**Comment by akh00** [comment]
PR has been changed according changes requested. Please review

**Comment by p0pr0ck5** [comment]
@akh00 please stop opening and closing this PR. Thanks.

File: kong/plugins/aws-lambda/handler.lua [hunk 19]
```diff
@@ -10,12 +10,13 @@ local cjson = require "cjson.safe"
...
 local string_find = string.find
+local string_fmt = string.format
 local ngx_req_get_headers = ngx.req.get_headers
 local ngx_req_read_body = ngx.req.read_body
 local ngx_req_get_uri_args = ngx.req.get_uri_args
 local ngx_req_get_body_data = ngx.req.get_body_data
-
-local CONTENT_TYPE = "content-type"
+local ngx_encode_base64 = ngx.encode_base64
+local var = ngx.var
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: Do not localize `ngx.var` in this manner, if anything, localize it in function scope, but never in a module. Same thing with `ngx.ctx`:

https://github.com/openresty/lua-nginx-module#ngxctx

> Because of the metamethod magic, never "local" the ngx.ctx table outside your Lua function scope on the Lua module level due to worker-level data sharing.

Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Good catch. Changed

**Comment by akh00** [comment]
@p0pr0ck5 > please stop opening and closing this PR. Thanks.
How else can I start build which failing randomly on different tests?

**Comment by thibaultcha** [comment]
> How else can I start build which failing randomly on different tests?

Ignore those, we'll take care of it.

**Comment by p0pr0ck5** [comment]
> How else can I start build which failing randomly on different tests?

You can also retrigger the build via the travis CI interface.

**Comment by akh00** [comment]
@p0pr0ck5 > You can also retrigger the build via the travis CI interface. 
I'm afraid I need to have "write" access to the repository for that

**Comment by NaveenGoli** [comment]
@thibaultcha  We prefer to get this change from official location as we plan to go to PROD with it and would like to avoid any merges in future.

**Comment by akh00** [comment]
Please review requested changes

File: kong/plugins/aws-lambda/schema.lua [hunk 20]
```diff
@@ -1,18 +1,22 @@
 return {
   fields = {
-    timeout = {type = "number", default = 60000, required = true },
-    keepalive = {type = "number", default = 60000, required = true },
-    aws_key = {type = "string", required = true},
-    aws_secret = {type = "string", required = true},
-    aws_region = {type = "string", required = true, enum = {
+    timeout = { type = "number", default = 60000, required = true },
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/schema.lua]: While we appreciate trying to keep style consistent, these whitespace changes here are undesired IMO. This dirties the commit history, as now _all_ the fields in the schema will have been changed with this PR, which does not accurately reflect history. A better approach would be to leave this whitespacing as it, commit your new fields in the proper style, and we can make a separate style cleanup commit in the future.

File: kong/plugins/aws-lambda/handler.lua [hunk 21]
```diff
@@ -25,32 +25,83 @@ function AWSLambdaHandler:new()
...
+  local body_json, err = cjson.encode(upstream_body)
+  if not body_json then
+    ngx.log(ngx.ERR, "[aws-lambda] could not encode JSON ",
+                   "body for forwarded request values: ", err)
+  end
 
-  local host = string.format("lambda.%s.amazonaws.com", conf.aws_region)
-  local path = string.format("/2015-03-31/functions/%s/invocations",
+  local host = string_fmt("lambda.%s.amazonaws.com", conf.aws_region)
+  local path = string_fmt("/2015-03-31/functions/%s/invocations",
```
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/handler.lua]: This is another case where this should be done in a separate commit, outside this PR. Indeed this is the correct usage, but we want to keep commit atomic; since this change is unrelated, it should go in a separate PR.

File: kong/plugins/aws-lambda/schema.lua (hunk 20 above)
Comment (akh00) [inline kong/plugins/aws-lambda/schema.lua]: It was explicit request from @thibaultcha. I would not ever change this because I agree that PR should be for single purpose.

File: kong/plugins/aws-lambda/handler.lua (hunk 21 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: It's another explicit request from @thibaultcha . If I remove those changes from PR may I have the request approved?

File: kong/plugins/aws-lambda/schema.lua (hunk 20 above)
Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/schema.lua]: I believe that request was only for the new code, not existing entries that don't meet style.

Comment (akh00) [inline kong/plugins/aws-lambda/schema.lua]: So it's okay when new code is in different style with the existing? I'm just trying to understand the pattern I should follow to pass the code review

Comment (p0pr0ck5) [inline kong/plugins/aws-lambda/schema.lua]: In order to maintain atomicity of the patchset, yet. Otherwise the size of your patchset if needlessly increased, and the git log does not accurately reflect the nature of the change.

Comment (thibaultcha) [inline kong/plugins/aws-lambda/schema.lua]: > It was explicit request from @thibaultcha. I would not ever change this because I agree that PR should be for single purpose.

No, please don't say such things. PR comments are always in the scope of said PR. I instructed you to update the code style of your change, and you went ahead and decided to update the entire table. That's undesired, but not my fault. Please propose the new line with the appropriate style, and leave the rest untouched.

**Comment by akh00** [comment]
Please review

**Comment by p0pr0ck5** [comment]
@akh00 please do not open and close this PR repeatedly. We will review this PR when we have the time and are able to; continuing to harass us is not going to speed the process along further. If you need this patch urgently for your own environment you are very welcome to include the patch in your own build. Thank you.

**Comment by akh00** [comment]
I'm closing and reopening request to run build which keeps failing in random places. I assume that green build is the "necessary check" for PR to be reviewed (at least it looks like this on this very page). I have no other means to rerun the build. I have absolutely no intention to harass anybody and will be just happy to have other way. I'm sorry for any inconveniences it brings to you.

**Comment by p0pr0ck5** [comment]
@akh00 this PR needs to be refactored following the merge of https://github.com/Mashape/kong/pull/2503.

**Comment by akh00** [comment]
Please review. I've refactored following the merge of #2503.

File: kong/plugins/aws-lambda/handler.lua [hunk 22]
```diff
@@ -20,19 +25,64 @@ function AWSLambdaHandler:new()
...
 
 local function retrieve_parameters()
-  ngx_req_read_body()
-
   return utils.table_merge(ngx_req_get_uri_args(), public_utils.get_body_args())
 end
 
+local function retrieve_body(content_type)
+  local raw_body = ngx_req_get_body_data()
+  if raw_body == nil then return nil end
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: style: we want to avoid 1 line branches. This should be:
```lua
if not raw_body then
  return nil
end
```

File: kong/plugins/aws-lambda/handler.lua [hunk 23]
```diff
@@ -20,19 +25,64 @@ function AWSLambdaHandler:new()
...
+
+  if content_type then
+    if string_find(content_type, "application/json", nil, true) then
+      local body_json, err = cjson.decode(raw_body)
+      if not body_json then
+          ngx.log(ngx.ERR, "[aws-lambda] could not encode JSON ",
+          "body for forwarded request values: ", err)
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: style: we need to align this second string with the first argument of this function:
```lua
ngx.log(ngx.ERR, "[aws-lambda] could not encode JSON ",
        "body for forwarded request values: ", err)
```

File: kong/plugins/aws-lambda/schema.lua [hunk 24]
```diff
@@ -14,5 +14,9 @@ return {
     log_type = {type = "string", required = true, default = "Tail", 
                        enum = {"Tail", "None"}},
     port = { type = "number", default = 443 },
+    forward_request_body = { type="boolean", default = false, required = true },
+    forward_http_headers = { type="boolean", default = false, required = true },
+    forward_http_method = { type="boolean", default = false, required = true },
+    forward_request_uri = { type="boolean", default = false, required = true },
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/schema.lua]: style: `type = "boolean"`

File: kong/plugins/aws-lambda/handler.lua [hunk 25]
```diff
@@ -20,19 +25,64 @@ function AWSLambdaHandler:new()
...
+  ngx_req_read_body()
+
+  local upstream_body
+
+  if conf.forward_request_body or conf.forward_http_headers
+     or conf.forward_http_method or conf.forward_request_uri then
+     local var = ngx.var
+     upstream_body = {
+       request_body_args = not conf.forward_request_body and retrieve_parameters()
+                           or ngx_req_get_uri_args(),
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: I'm not sure I understand the value in having both `request_body` and `request_body_args`. Since the point of parsing the body is to manipulate it in the Lua land (inside Kong), but once it's forwarded upstream, `request_body` should be enough, and the labda function can parse it itself.

As of now, if you have a JSON body, then both `request_body_args` and `request_body` would have the same value.

File: kong/plugins/aws-lambda/handler.lua [hunk 26]
```diff
@@ -20,19 +25,64 @@ function AWSLambdaHandler:new()
...
+       request_body = conf.forward_request_body and retrieve_body(var.http_content_type) or nil,
+       request_http_headers = conf.forward_http_headers and ngx_req_get_headers() or nil,
+       request_http_method = conf.forward_http_method and var.request_method or nil,
+       request_uri = conf.forward_request_uri and var.request_uri or nil
+     }
+  else
+     upstream_body = retrieve_parameters()
+  end
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: It seems like this is backwards compatible enough. Sadly, we force the upstream request body to be JSON, instead of forwarding it as-is (or I should say "at all", because this should be optional as there is a memory impact on large bodies) but well. Backwards-compatible is good enough.

File: kong/plugins/aws-lambda/handler.lua [hunk 27]
```diff
@@ -20,19 +25,64 @@ function AWSLambdaHandler:new()
...
+
+  local body_json, err = cjson.encode(upstream_body)
+  if not body_json then
+    ngx.log(ngx.ERR, "[aws-lambda] could not encode JSON ",
+                   "body for forwarded request values: ", err)
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: style: needs to be aligned like the other `ngx.log` call previously mentioned.

File: kong/plugins/aws-lambda/schema.lua [hunk 28]
```diff
@@ -14,5 +14,9 @@ return {
     log_type = {type = "string", required = true, default = "Tail", 
                        enum = {"Tail", "None"}},
     port = { type = "number", default = 443 },
```
Comment (thibaultcha) [inline kong/plugins/aws-lambda/schema.lua]: I think this should still be named `aws_port`.

File: kong/plugins/aws-lambda/handler.lua (hunk 17 above)
Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: still

File: kong/plugins/aws-lambda/handler.lua (hunk 25 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: No If "request_body" is "true" request_body_args is going to include only uri parameters but not body itself. In case of "forward_request_body" is true body is forwarded only in request_body. The difference for Content-type json is that the body is parsed. Which repeats the existing style - parsing json for "body arguments".

Comment (thibaultcha) [inline kong/plugins/aws-lambda/handler.lua]: Why would a variable containing request bodies contain URI arguments? That is utterly confusing.

We should forward plain bodiesand plain URI, without any need to parse it on our side.

File: kong/plugins/aws-lambda/handler.lua (hunk 26 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: No we are forwarding upstream body as json if it's json and we are forwarding it as it is in other case when we are instructed to.

File: kong/plugins/aws-lambda/handler.lua (hunk 25 above)
Comment (akh00) [inline kong/plugins/aws-lambda/handler.lua]: Because it's how it works now - body arguments include URI parameters. Moreover currently if body is json then URI parameters are mixed to the body as additional fields (potentially overwriting existing).
If it's the only concern I'll just remove this logic in hope to get PR approved

**Comment by akh00** [comment]
Please review

**Comment by akh00** [comment]
@thibaultcha @p0pr0ck5  could you please review. Tests keep failing randomly. I don't want to close/reopen PR to rerun tests

**Comment by p0pr0ck5** [comment]
@akh00 we have already asked you multiple times to stop opening/closing this PR. Please stop doing so. We will review this PR when we are able to.

**Comment by akh00** [comment]
I've resolved conflicts with last merged commit. Please review

**Comment by NaveenGoli** [comment]
when can this be merged after conflict resolution?


Thanks

Naveen



________________________________

**Comment by thibaultcha** [comment]
I will take care of it when my plate is clean. So far the current implementation looks ok, with a couple of minor adjustments to be taken care of at merge-time.

**Comment by akh00** [comment]
Sorry for bothering. Any ETA for merging it?

**Comment by NaveenGoli** [comment]
any luck? we are eagerly waiting for this fix and want to avoid a private build

**Comment by Mentis** [comment]
We also need this fix. Any updates on when this will be available?

**Comment by saulmadi** [comment]
We also need this fix. Any updates on when this will be available?

**Comment by numidiasoft** [comment]
We also need this fix. Any updates on when this will be available?

**Comment by erran** [comment]
We're currently using a patched version of the Lambda plugin I built in production, at work, to include changes that haven't been released yet (even though they've been merged). You can take the `aws-lambda` folder from this pull request and package/install it — https://getkong.org/docs/0.10.x/plugin-development/distribution/

You can update the plugin name in the folder path, require statements, and plugin initialise function to support both this version the mainstream version at the same time.

Even when this is merged it may be a while before it's released.

@akh00 or anyone else who needs it could package it as a luarock for easier distribution as a custom plugin as well.

**Comment by thibaultcha** [comment]
@akh00 I just opened #2823 which is a reworked version of your patch, rebased on the most recent master and fixing quiet a few issues. It changes a few things, especially regarding the interpretation of the set `forward_request_*` values when crafting the upstream request body. Will you have a look at it to confirm it solves your use-case?

Thanks, I'll be closing this in the meanwhile!
